server/user.js

```diff
@@ -0,0 +1,182 @@
+/**
+ * @fileoverview user function handler
+ */
+var fs = require('fs');
+var assert = require('assert');
+
+var log = require('./log.js');
+var utils = require('./utils.js');
+
+/** @type {user} */
+module.exports = user;
+
+/**
+ * @constructor
+ */
+function user() {}
+
+/** @enum {string} */
+user.QueryType = {
+  SIGNUP: 'sign-up',
+  SIGNIN: 'sign-in'
+};
+
+/**
+ * Expire time of cookie.
+ * @const {number}
+ */
+user.cookieExpireTime = 24 * 60 * 60 * 1000;
+
+/**
+ * @typedef {{
+ *   email: (string|undefined),
+ *   username: string,
+ *   password: string,
+ *   confirmed: (boolean|undefined)
+ * }}
+ */
+user.Info;
+
+/**
+ * @typedef {{
+ *   error: string
+ * }}
+ */
+user.Error;
+
+/**
+ * Checks the user information for signing in.
+ * @param {!mongodb.Db} db Session database.
+ * @param {string} userPath Directory of user information file.
+ * @param {!user.Info} userInfo User Information
+ * @param {function(!Object)} callback Callback function.
+ */
+user.signUp = function(db, userPath, userInfo, callback) {
+  var checkDuplicate = {
+    duplicated: false,
+    elements: []
+  };
+  var cursor = db.collection('userInfo').find({
+    $or: [{username: userInfo.username}, {email: userInfo.email}]});
+  var data = [];
+  cursor.each(function(err, doc) {
+    assert.equal(err, null, 'error occurs');
+    if (doc != null) {
+      data.push(doc);
+    } else {
+      var result = authenticateCallback();
+      callback(result);
+    }
+  });
+  var authenticateCallback = function() {
+    if (data.length) {
+      checkDuplicate.duplicated = true;
+      console.log(data);
+      data.forEach(function(item) {
+        console.log(item);
+        if (item.email == userInfo.email) {
+          checkDuplicate.elements.push('email: ' + userInfo.email);
+        }
+        if (item.username == userInfo.username) {
+          checkDuplicate.elements.push('username: ' + userInfo.username);
+        }
+      });
+    }
+    if (checkDuplicate.duplicated) {
+      var errorMessage = checkDuplicate.elements.join(' and ') + ' exist';
+      errorMessage += checkDuplicate.elements.length == 1 ? 's' : '';
+      return {
+        error: errorMessage
+      };
+    } else {
+      db.collection('userInfo').insertOne(userInfo);
+      log.serverLog('user information saved');
+      var folder = userPath + userInfo.username + '/';
```

Comment:
Then we will let the log code rely on that the user folder exists, which is risky and error prone assuming there are unexpected errors or inconsistency somewhere. For example, if db insertOne was successful, and log.serverLog failed because of log disk full :/. Then the server panics and the user folder won't be created. But the user was created successfully in db, and later when you write to logs you don't have the folder.

Of course the above is a rare scenario and I'm being overly cautious here. But it is kind of a style that we "create directory on demand". A rule of thumb is that each code does those checks that are related to its primary tasks. Therefore "whoever need to write create the file/directory" IMO makes things easier for debugging.

In this case, the signup function already spends a lot of its work updating the db. It seems to me that separating the folder creation makes its task lighter and cleaner (save one argument for logPath).